### ISSUE TITLE
Performance issue with azure batch

### ISSUE THREAD

**Issue by andrew-vdb** [Azure/Batch#58]
First Task
![image](https://user-images.githubusercontent.com/12871379/57010828-171b2c00-6bff-11e9-8a68-18f66b4502ba.png)

![image](https://user-images.githubusercontent.com/12871379/57010887-534e8c80-6bff-11e9-9abc-9f1b0b3d4620.png)

Second Task
![image](https://user-images.githubusercontent.com/12871379/57010838-213d2a80-6bff-11e9-8b09-ded891301bd8.png)

![image](https://user-images.githubusercontent.com/12871379/57010899-61041200-6bff-11e9-8dfc-970b0547ce03.png)

Second task has dependency with first task.

I have task A and task B,
task B depended on task A

task A run for 1 second
task B run for 0 second

If you take a look to the timing to execute these 2 tasks, it is seriously slow

Between 10:47:44 and 10:47:51, I lose 7 seconds of doing nothing?

**Comment by alfpark** [comment]
I think the delay is due to dependency processing. @theunrepentantgeek might be able to provide more insight here.

**Comment by andrew-vdb** [comment]
This is serious performance issue, can this be fixed or should I take a look another solution?

**Comment by andrew-vdb** [comment]
Losing 4 seconds between 10:47:38 - 10:47:42, also quite long for only downloading resources

**Comment by theunrepentantgeek** [comment]
It's difficult to assess exactly what's going on without more information; here's my best guess.

I suspect that the majority of the 7 second delay you're seeing is due to normal scheduling latency. Batch is optimized for tasks that take a few minutes to a few days of execution time. For these longer running tasks, a delay of a few seconds isn't usually considered significant.

Our scheduler is somewhat adaptive. If there are no tasks available to schedule, it backs off, checking less frequently. This can increase the time it takes for a new task to be picked up . I believe the upper limit is 10 seconds. Once a task has been found, however, it will aggressively schedule all available tasks while there are compute nodes available.

To illustrate, I created a pool with 10 compute nodes and then ran a small job with 10 tasks. Tasks 2-10 were all dependent on task 1. 

| Task    | Start                   | Finish                  |
| ------- | ----------------------- | ----------------------- |
| task-1  | 2019-05-01T23:05:11.252 | 2019-05-01T23:05:45.231 |
| task-2  | 2019-05-01T23:05:51.319 | 2019-05-01T23:06:26.292 |
| task-3  | 2019-05-01T23:05:51.317 | 2019-05-01T23:06:26.297 |
| task-4  | 2019-05-01T23:05:51.319 | 2019-05-01T23:06:24.184 |
| task-5  | 2019-05-01T23:05:51.315 | 2019-05-01T23:06:26.287 |
| task-6  | 2019-05-01T23:05:51.316 | 2019-05-01T23:06:26.292 |
| task-7  | 2019-05-01T23:05:51.320 | 2019-05-01T23:06:26.301 |
| task-8  | 2019-05-01T23:05:51.322 | 2019-05-01T23:06:26.293 |
| task-9  | 2019-05-01T23:05:51.316 | 2019-05-01T23:06:26.347 |
| task-10 | 2019-05-01T23:05:51.317 | 2019-05-01T23:06:29.384 |

You can see there's a delay of just over 6 seconds between the end of task-1, and the start of the other 9 tasks. They were all started within 7 milliseconds of each other.

For high throughput jobs with dozens or thousands of tasks, we usually see tasks dispatched to available compute nodes in a second or less.

To pick up on Fred's comment above, our dependency processor releases followup tasks once their dependencies have been satisfied. It doesn't do the scheduling directly, however: it places those tasks on the same queue as other tasks.

If you'd like someone to investigate further, please create a support issue via the Azure Portal. We'll need to know the your batch account name & region, the pool id, job name, and accurate timestamps for your issue.

If you'd like to discuss your scenario and get some advice on using Batch effectively, you could email the Ask Azure Batch (askwabatch@microsoft.com) mailing list.

**Comment by andrew-vdb** [comment]
I already submitted support request

I have 1 node, 1 job, max 2 task running on the same time
I never delete the job (should not matter right)
I create 3 set of tasks (See creation time)
Each set have 2 tasks, task A and task B

![image](https://user-images.githubusercontent.com/12871379/57063522-d76a4800-6cc3-11e9-9064-2c4aa634b150.png)

Between 8:53:38 - 8:53:44, there is nothing running and I already lose 6 seconds.

If this something that can not be fixed, then I will do my work around....
The documentation should be updated too, be aware of delay between task A and task B if you use task dependency.

> To pick up on Fred's comment above, our dependency processor releases followup tasks once their dependencies have been satisfied. It doesn't do the scheduling directly, however: it places those tasks on the same queue as other tasks.

Its not clear why batch idle between 8:53:34 - 8:53:37 and between 8:53:38 - 8:53:44
Don't forget we do pay for this idle time.

**Comment by andrew-vdb** [comment]
I do my workaround, no task dependency anymore,
I only care about **how task can be started asap now**
Task created 12:51:08
Started at 12:51:18
10 seconds delay even I already warm up the task scheduler (by running task before it)

![image](https://user-images.githubusercontent.com/12871379/57071175-e6f48b80-6cd9-11e9-8203-305edcc02e35.png)

![image](https://user-images.githubusercontent.com/12871379/57071193-f673d480-6cd9-11e9-84c2-855cf7ebf95d.png)

Putting more tasks seems improve the start time, now its 4 seconds
![image](https://user-images.githubusercontent.com/12871379/57071725-a72ea380-6cdb-11e9-8fe1-1475daaee41a.png)

![image](https://user-images.githubusercontent.com/12871379/57071752-b3b2fc00-6cdb-11e9-9c85-18f43c0b515b.png)


Can we adjust the upper limit 10 seconds to 1 second?

**Comment by bgklein** [comment]
@andrew-vdb  One thing which may be contributing to this is that you are using 'Task Default User' rather than 'Pool Default User'. If you are running on Windows, user creation is a rather heavy weight operation and by having this at the task level you incurring it in the tasks preparing step for each task. If you set it to pool then you will only incur the user creation time once at pool creation and all tasks will run under the same user profile.  

This is only one piece of the timing but thought I'd chime in just in case you don't need user isolated tasks and were running on Windows.

**Comment by andrew-vdb** [comment]
@bgklein I have stuff running in linux container too, i lose 4-5 seconds
![image](https://user-images.githubusercontent.com/12871379/57123236-6721fc00-6d81-11e9-91c3-bbdf47eab7d1.png)
![image](https://user-images.githubusercontent.com/12871379/57123249-6f7a3700-6d81-11e9-99f4-23fd3f3a2f80.png)

**Comment by alfpark** [comment]
Typical task scheduling latencies are on the order of 2-5 seconds in the pool user case. This is variable and may be longer due to various factors such as back off, queue depth, and other lower layer dependencies that may be exhibiting longer latencies than usual.

If you need scheduling latencies that are on the order of less than 1 second, Batch will probably not be able to fit your requirements. As @theunrepentantgeek has pointed out, Batch is designed to target workloads where scheduling latency on the order of a few seconds is not in the critical path.
